compiler/vm/cbackend2.nim

```diff
@@ -0,0 +1,1097 @@
+
+  let ast =
+    if module.stmts.len > 1: newTree(nkStmtList, module.stmts)
+    elif module.stmts.len == 1: module.stmts[0]
+    else: newNode(nkEmpty)
+
+  let tn = transformStmt(c.graph, c.idgen, c.module, ast)
+  let r = c.genStmt(tn)
+
+  if unlikely(r.isErr):
+    config.localReport(r.takeErr)
+
+  c.endProc()
+
+  if c.irs.len > 2:
+    # call to ``nimTestErrorFlag`` at the very end of every 'init' procedure
+    # in order to report any unhandled exception
+    discard c.irs.irCall(
+      c.irs.irProc(c.passEnv.getCompilerProc("nimTestErrorFlag")))
+
+  # if the resulting procedure is empty, `irs` still has two items: the
+  # 'join's at the end
+  # TODO: use a more forward-compatible approach for testing if there's no
+  #       code
+  result =
+    if c.irs.len > 2: some(move c.irs)
+    else:             none(IrStore3)
+
+proc generateCodeForProc(c: var TCtx, s: PSym): IrGenResult =
+  assert s != nil
+  #debugEcho s.name.s, "(", s.kind, "): ", c.config.toFileLineCol(s.info)
+  let body = transformBody(c.graph, c.idgen, s, cache = false)
+  c.irs.reset()
+  c.options = s.options
+  result = genProc(c, s, body)
+
+proc unwrap[T](c: TCtx, r: Result[T, SemReport]): T =
+  if r.isErr:
+    c.config.localReport(r.takeErr)
+  else:
+    result = r.unsafeGet
+
+proc generateGlobalInit(c: var TCtx, f: var CodeFragment, defs: openArray[PNode]) =
+  ## Generates and emits code for the given `{.global.}` initialization
+  ## statements (`nkIdentDefs` in this case) into `f`
+  template swapState() =
+    #swap(c.code, f.code)
+    #swap(c.debug, f.debug)
+    swap(c.prc, f.prc)
+
+  # In order to generate code into the fragment, the fragment's state is
+  # swapped with the `TCtx`'s one
+  swapState()
+
+  for def in defs.items:
+    assert def.kind == nkIdentDefs
+    for i in 0..<def.len-2:
+      # note: don't transform the expressions here; they already were, during
+      # transformation of their owning procs
+      let
+        a = c.genExpr(def[i])
+        b = c.genExpr(def[^1])
+        r = c.irs.irAsgn(askInit, c.unwrap a, c.unwrap b)
+
+  # Swap back once done
+  swapState()
+
+proc generateEntryProc(g: PassEnv, mlist: ModuleList,
+                       modules: seq[ModuleDataExtra]): IrStore3 =
+  ## Generates and returns the code for the entry procedure. The entry
+  ## procedure is responsible for calling the data-init and init procedures
+  ## of each module. On exit, the entry procedure returns the value of the
+  ## ``programResult`` global
+
+  # first, emit calls to the modules' data-init procedure
+  for id in mlist.modulesClosed.items:
+    let m {.cursor.} = modules[id]
+    if m.dataInitProc != NoneProc:
+      discard result.irCall(result.irProc(m.dataInitProc))
+
+    if sfSystemModule in m.flags:
+      # the 'init' procedure for the ``system`` module must be called right
+      # after the data init procedure
+      # XXX: that's what ``cgen`` does, but what's the reason behind this
+      #      requirement?
```

Comment:
The question was in the direction of why the `init` procedure for `system` is called before the `DatInit` procedure of other modules.

Your assessment is correct, it's necessary because the `DatInit` procedures are also responsible for loading used dynamic libraries and for setting up the procedures/globals depending on them, which is something that requires various `system` parts to be initialized (`cstderr`, the GC/allocator, etc.).